docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,529 @@
+- Start Date: 2021-02-08
+- FLIP PR: [#1011](https://github.com/google/flax/pull/1011)
+- FLIP Issue: [#1009](https://github.com/google/flax/issues/1009)
+
+Table of contents:
+
+- [Summary]
+- [Motivation]
+- [Functional API]
+  - [Functional Implementation]
+  - [Using Optax]
+  - [Multi Optimizer]
+  - [Functional Usage]
+  - [Linen Helper]
+- [Previous API]
+  - [Previous Implementation]
+  - [Previous Usage]
+- [Update Plan]
+
+
+# Summary
+[Summary]: #summary
+
+Our current API (referred to as [previous API] in this document) uses a pattern
+where an `Optimizer` dataclass is created from a pytree of `target` variables
+and from an `OptimizerDef` that defines how to update optimizer state,
+hyperparameters, and target variables. This pattern is relatively complex for
+implementing a simple optimizer, while being quite verbose in the typical Linen
+train step (especially when using mutable state collections).
+
+This FLIP proposes a new purely [functional API] where optimizers are simply
+defined by an `init()` and an `update()` function. This new pattern allows
+[using Optax] optimizers that follow a very similar functional pattern.
+
+In the end, this document also proposes a new small [Linen helper] that takes
+care of the usual splitting and merging of variables for optimization.
+
+
+# Motivation
+[motivation]: #motivation
+
+Flax contains some optimizers in the `flax.optim` package, but that list is far
+from exhaustive and ideally we would instead use JAX optimizers from a dedicated
+PyPi package.
+
+Deepmind's [Optax] library is such a dedicated package that already implements a
+choice of interesting optimizers and it also provides a framework to compose new
+optimizers from reusable gradient transformations.
```

Comment:
Small suggestion:

- Nit/minor refactoring

```suggestion
DeepMind already has a dedicated library — [Optax] — that already implements a
wide range of interesting optimizers and provides a framework to compose new
optimizers from reusable gradient transformations.
```

Reply:
Thanks for the suggestion - updated.

Reply:
No worries, glad to be of service